<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Загрузим-данные-и-изучим-общую-информацию" data-toc-modified-id="Загрузим-данные-и-изучим-общую-информацию-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Загрузим данные и изучим общую информацию</a></span><ul class="toc-item"><li><span><a href="#Промежуточный-вывод-предварительного-исследования-данных" data-toc-modified-id="Промежуточный-вывод-предварительного-исследования-данных-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Промежуточный вывод предварительного исследования данных</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Исследование-аномалий" data-toc-modified-id="Исследование-аномалий-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Исследование аномалий</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li></ul></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Категоризация данных</a></span><ul class="toc-item"><li><span><a href="#Категоризация-по-уровню-доходов" data-toc-modified-id="Категоризация-по-уровню-доходов-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Категоризация по уровню доходов</a></span></li><li><span><a href="#Категоризация-по-цели-взятия-кредита" data-toc-modified-id="Категоризация-по-цели-взятия-кредита-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Категоризация по цели взятия кредита</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы" data-toc-modified-id="Ответы-на-вопросы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Приведите-возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="Приведите-возможные-причины-появления-пропусков-в-исходных-данных.-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Приведите возможные причины появления пропусков в исходных данных.</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-6.6"><span class="toc-item-num">6.6&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></li></ul></div>

# Описание проекта


Входными данными для исследования является информация о клиентах банка, переданная кредитным отделом.    
Основной целью проекта является нахождение взаимосвязи между количеством детей клиента, его семейным положением, уровнем дохода и целей кредита на факт возвращения кредита в срок. 

В работе постараемся ответить на следующие вопросы:  
- Есть ли зависимость между количеством детей и возвратом кредита в срок?  
- Есть ли зависимость между семейным положением и возвратом кредита в срок?  
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?  
- Как разные цели кредита влияют на его возврат в срок?  
- Приведите возможные причины появления пропусков в исходных данных.  

# Описание данных

`children` — количество детей в семье  
`days_employed` — общий трудовой стаж в днях  
`dob_years` — возраст клиента в годах  
`education` — уровень образования клиента  
`education_id` — идентификатор уровня образования  
`family_status` — семейное положение  
`family_status_id` — идентификатор семейного положения  
`gender` — пол клиента  
`income_type` — тип занятости  
`debt` — имел ли задолженность по возврату кредитов  
`total_income` — ежемесячный доход  
`purpose` — цель получения кредита  

# Загрузим данные и изучим общую информацию

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

import warnings

Отключим предупреждения.

In [2]:
warnings.filterwarnings('ignore')

In [3]:
# считаем данные из файла csv
data = pd.read_csv(
    r'C:\Users\Dell Inspirion 15 #1\JyputerRoot\Проекты по Практикуму\Обработанные проекты\Datasets\02_заёмщики.csv')

In [4]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [5]:
data.shape

(21525, 12)

In [6]:
data.describe().T

count           mean            std           min  \
children          21525.0       0.538908       1.381587     -1.000000   
days_employed     19351.0   63046.497661  140827.311974 -18388.949901   
dob_years         21525.0      43.293380      12.574584      0.000000   
education_id      21525.0       0.817236       0.548138      0.000000   
family_status_id  21525.0       0.972544       1.420324      0.000000   
debt              21525.0       0.080883       0.272661      0.000000   
total_income      19351.0  167422.302208  102971.566448  20667.263793   

                            25%            50%            75%           max  
children               0.000000       0.000000       1.000000  2.000000e+01  
days_employed      -2747.423625   -1203.369529    -291.095954  4.017554e+05  
dob_years             33.000000      42.000000      53.000000  7.500000e+01  
education_id           1.000000       1.000000       1.000000  4.000000e+00  
family_status_id       0.000000       0.000000       1.000000  4.000000e+00  
debt                   0.000000       0.000000       0.000000  1.000000e+00  
total_income      103053.152913  145017.937533  203435.067663  2.265604e+06

In [7]:
data.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 11.3 MB


## Промежуточный вывод предварительного исследования данных

Из предварительного анализа датасета мы видим, что имеются пропуски в колонках с данными об общем трудовом стаже и ежемесячным доходе, эти пропуски планируется заполнить медианными значениями для соответсвующих категорий данных.  

В рамках данного проекта информацию о днях общей занятости и ежемесячном доходе клиента достаточно получить с точностью до целого значения, поэтому изменим тип данных float на int там, где это целесообразно после заполнения пропусков.

# Предобработка данных

## Заполнение пропусков

Найдем количество пропущенных значений для каждого столбца.

In [8]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Столбец `days_employed`, содержащий информацию об общем трудовом стаже в днях имеет 2174 пропуска, такое же количество пропусков имеет столбец `total_income`. Возможно, что пропуски в обоих столбцах обусловлены одной ошибкой.  

Проверим, имеют ли клиеты с пропусками данных в этих столбцах, информацию об источнике дохода.

In [9]:
data[data['total_income'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Клиенты без информации об ежемесячном доходе и общем трудовом стаже имеют информацию об источнике дохода.

Было принято пропуски в столбце с ежемесячным доходом заполнить медианным значением по типу занятости клиента.

In [10]:
# Заполним пропуски в столбце 'total_income' медианными значениями
data.loc[data['total_income'].isna(), 'total_income'] = data.groupby(
    'income_type')['total_income'].transform('median')

In [11]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

Заполним пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.

In [12]:
# Заполним пропуски в столбце 'days_employed' медианными значениями
data.loc[data['days_employed'].isna(), 'days_employed'] = data.groupby(
    'income_type')['days_employed'].transform('median')

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

## Изменение типов данных

In [14]:
data[['days_employed', 'total_income']] = data[[
    'days_employed', 'total_income']].astype('int')

In [15]:
# проверим изменения
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int32 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int32 
 11  purpose           21525 non-null  object
dtypes: int32(2), int64(5), object(5)
memory usage: 1.8+ MB


## Исследование аномалий

Выведем перечень уникальных значений столбца `children`

In [16]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

Количество детей не может быть равно отрицательному значению, так же 20 детей в семье является аномальным значением. Найдем количество анамальных значений.

In [17]:
print('Количество аномалий в children:', len(
    data.query('children==20  or children==-1')))
print('Доля аномалий в children от общего количества значений:', len(
    data.query('children==20  or children==-1'))/len(data)*100, '%')

Количество аномалий в children: 123
Доля аномалий в children от общего количества значений: 0.5714285714285714 %


Удалим строки с аномальными значениями в столбце `children`

In [18]:
data = data.query('children!=-1 and children!=20')
data.head()

children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   Среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [19]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

## Обработка дубликатов

Проверим количество явных дубликатов в датасете.

In [20]:
data.duplicated().sum()

54

**Обработаем неявные дубликаты в столбце `education`**

In [21]:
data['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Заменим значения столбца на значения в нижнем регистре.

In [22]:
data['education'] = data['education'].str.lower()

In [23]:
data['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

**Аналогично для столбцов `income_type`, `gender`**

In [24]:
data['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

В столбце `income_type` нет неявных дубликатов.

In [25]:
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [26]:
data.groupby('gender')['children'].count()

gender
F      14154
M       7247
XNA        1
Name: children, dtype: int64

Удалим строку с некорректным гендером.

In [27]:
data = data.query("gender!='XNA'")

# Категоризация данных

## Категоризация по уровню доходов

На основании указанных ниже диапазонов создадим столбец `total_income_category` c категориями:  

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [28]:
# создадим фунцию categorize_income()

def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [29]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

## Категоризация по цели взятия кредита

In [30]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [31]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [32]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data.sample(10)

children  days_employed  dob_years education  education_id  \
3494          0         -10186         44   среднее             1   
11090         0          -1574         33   среднее             1   
13009         2          -3577         28   среднее             1   
2601          0           -879         27    высшее             0   
16479         0          -2674         61    высшее             0   
9020          0          -2699         35   среднее             1   
13904         0           -234         57   среднее             1   
2225          2          -3862         34   среднее             1   
9124          0           -282         48   среднее             1   
12415         3          -1460         37   среднее             1   

               family_status  family_status_id gender income_type  debt  \
3494         женат / замужем                 0      F   сотрудник     0   
11090       гражданский брак                 1      M   сотрудник     0   
13009        женат / замужем                 0      M   сотрудник     0   
2601   Не женат / не замужем                 4      F   сотрудник     0   
16479  Не женат / не замужем                 4      M   сотрудник     0   
9020   Не женат / не замужем                 4      F   сотрудник     0   
13904        женат / замужем                 0      F   компаньон     0   
2225         женат / замужем                 0      F   сотрудник     0   
9124               в разводе                 3      F   компаньон     0   
12415       гражданский брак                 1      M   сотрудник     0   

       total_income                                 purpose  \
3494         175451        строительство жилой недвижимости   
11090        142594        сделка с подержанным автомобилем   
13009        192592              покупка жилой недвижимости   
2601         118850                                   жилье   
16479        113759            на покупку своего автомобиля   
9020         114625                 покупка жилья для семьи   
13904         95462       покупка коммерческой недвижимости   
2225         140347  строительство собственной недвижимости   
9124         284275        строительство жилой недвижимости   
12415        227487                   на проведение свадьбы   

      total_income_category          purpose_category  
3494                      C  операции с недвижимостью  
11090                     C    операции с автомобилем  
13009                     C  операции с недвижимостью  
2601                      C  операции с недвижимостью  
16479                     C    операции с автомобилем  
9020                      C  операции с недвижимостью  
13904                     C  операции с недвижимостью  
2225                      C  операции с недвижимостью  
9124                      B  операции с недвижимостью  
12415                     B        проведение свадьбы

# Ответы на вопросы

## Есть ли зависимость между количеством детей и возвратом кредита в срок?

Столбец `debt` показывает имелась ли задолженность по кредитам. Сгруппировав значения по количеству детей, функция mean покажет процент невозрата кредитов.

In [33]:
children_and_debt = (data
                     .pivot_table(index='children', values='debt', aggfunc='mean')
                     .sort_values(by='debt')
                     .rename(columns={'debt': 'debt_%'})
                     )
children_and_debt['debt_%'] = round(children_and_debt['debt_%']*100, 2)
children_and_debt

debt_%
children        
5           0.00
0           7.51
3           8.18
1           9.22
2           9.44
4           9.76

In [34]:
data.groupby('children')['children'].count()

children
0    14148
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

**Вывод:** Проанализировав полученные данные можно сказать, что чем меньше детей у клиентов, тем реже они берут кредит. Большой разницы по проценту невозврата не наблюдается.  

Клиенты без детей в 7.5% случаев невозвращают деньги в долг.  

Клиенты с одним или двумя детьми не возвращают кредиты в 9% процентов случаев в среднем.  

У клиентов с тремя детьмя возникают трудности с просрочкой в 8% случаев.

Клиентов с четырмя детьми всего 41 и не возвращают вовремя кредиты они почти в 10% случаев, но данный показатель можно считать приблизительным, так как семьи с четырьмя детьми встречатся достаточно редко.

Клиентов с пятью детьми всего 9 и все они возвращали кредиты строго в срок. Но нельзя утверждать, что они самые надежные заемщики, так как для полноценного анализа недостаточно такой малой выборки.

## Есть ли зависимость между семейным положением и возвратом кредита в срок?

Аналогично предыдущему пункту найдем процент невозврата кредитов с группировкой по семейному положению.

In [35]:
debt_and_family = (data
                   .pivot_table(index='family_status', values='debt', aggfunc='mean')
                   .sort_values(by='debt')
                   .rename(columns={'debt': 'debt_%'})
                   )
debt_and_family['debt_%'] = round(debt_and_family['debt_%']*100, 2)
debt_and_family

debt_%
family_status                
вдовец / вдова           6.62
в разводе                7.06
женат / замужем          7.54
гражданский брак         9.26
Не женат / не замужем    9.75

In [36]:
data.groupby('family_status')['children'].count()

family_status
Не женат / не замужем     2799
в разводе                 1189
вдовец / вдова             952
гражданский брак          4159
женат / замужем          12302
Name: children, dtype: int64

Вывод: Анализ показал, что замужние/женатые клиенты в значительной степени чаще пользуются кредитами, т.к. у молодых семей как правило больше статей расхода. При этом процент невозврата составляет 7.5%.

Немного менее ответственными являются пары в гражданском браке, они не возвращают кредиты в срок в 9% случаев, что сродни показателям неженатых/незамужних клиентов.

Клиенты разводе не возвращают деньги вовремя в 7% случаев, а вдовцы в 6.6%.

Нельзя однозначно утверждать, что семейный статус в значительной степени влияет на уровень надежности заемщика.

## Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Для поиска необходимой нам зависимости воспользуемся проведенной ранее категоризацией  значений из столбца "total_income". Напомним, что категоризация проводилась по следующим критериям:
    
- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [37]:
income_and_debt = (data
                   .pivot_table(index='total_income_category', values='debt', aggfunc='mean')
                   .sort_values(by='debt')
                   .rename(columns={'debt': 'debt_%'})
                   )
income_and_debt['debt_%'] = round(income_and_debt['debt_%']*100, 2)
income_and_debt

debt_%
total_income_category        
D                        6.02
B                        7.06
A                        8.00
C                        8.46
E                        9.09

In [38]:
data.groupby('total_income_category')['children'].count().sort_values()

total_income_category
E       22
A       25
D      349
B     5013
C    15992
Name: children, dtype: int64

Вывод:Как мы видим из таблицы: люди с низким доходом (категорая Е 0-30 000) и люди с высоким доходом (категория А с оходом свыше миллиона) редко пользуются кредитами. Выборка настолько невелика, что её можно исключить из анализа надёжности клиента в зависимости от уровня доходов.
    
Самой многочисленной группой, которая пользуется кредитами, является категория С (доход 50 000 - 200 000), они же являются наиболее ненадежными заемщиками, процент невозврата кредита в срок равен почти 8,5%.

## Как разные цели кредита влияют на его возврат в срок?

Аналогично предыдущему пункту, найдем процент невозврата кредита в зависимости от его цели.

In [40]:
purpose_and_debt = (data
                    .pivot_table(index='purpose_category', values='debt', aggfunc='mean')
                    .sort_values(by='debt')
                    .rename(columns={'debt': 'debt_%'})
                    )
purpose_and_debt['debt_%'] = round(purpose_and_debt['debt_%']*100, 2)
purpose_and_debt

debt_%
purpose_category                
операции с недвижимостью    7.24
проведение свадьбы          7.83
получение образования       9.23
операции с автомобилем      9.33

Вывод: В категориях "операции с недвижимостью" и "проведение свадьбы" не возвращается в срок 7-8 процентов кредитов. В категориях "операции с автомобилем" и "получение образования" в срок не возвращиется примерно 9 процентов. Можно сказать, что невозврат кредита в срок не сильно зависит от цели его получения при всех прочих равных.

## Приведите возможные причины появления пропусков в исходных данных.

Ответ: Пропуски изначально мы наблюдали в столбцах с общим трудовым стажем и ежемесячным доходом. Возможно, что пропуски появились из-за ошибки выгрузки, либо ошибки оператора, который эту информаци вносил, так как в строках с пропущенными данными иногда имеется информация о количестве просроченных кредитов. То есть мы знаем, что клиент уже не возвращал деньги в срок, но не знаем его доходов и трудового стажа. Так же причиной может быть то, что база по клиентам получена частично от другого банка или других источников, тогда мы можем знать что кредиты он уже не возвращал, но в нашем банке мог не брать, например.

## Общий вывод

Результаты исследования не показали значительной корреляции между возратом кредита в срок и семейным положением, количеством детей или уровнем дохода.  

При среднем значении невозврата в срок кредита равным 8%, самыми ответственными являются:

Клиенты без детей - 7.5% случаев невозврата кредита в срок.  
Вдовцы - 6.6 % случаев невозврата кредита в срок.  
Клиенты с доходом 30 000 - 50 000 - 6% случаев невозврата кредита в срок.  
И клиенты берущие кредит на операции с недвижимостью - 7.2% случаев невозврата кредита в срок.    

При этом самые злостные неплательщики:

Клиенты с одним или двумя детьми - 9.3 % случаев невозврата кредита в срок в среднем.  
Находящиеся в гражданском браке или не женатые - 9.5% случаев невозврата кредита в срок в среднем.  
С уровнем дохода 50 000 - 200 000 - 8.5% случаев невозврата кредита в срок.  
Берущие кредит на операции с автомобилем - 9.3% случаев невозврата кредита в срок.  

Однозначно можно сказать, что чаще всего кредитами пользуятся:
- клиенты без детей;
- женатые/замужние;
- с уровнем дохода 50 000 - 200 000;
- на операции с недвижимостью.

Реже всего берут кредит:
- клиенты с большим количеством детей;
- вдовы/вдовцы или клиенты в разводе;
- с очень большим или маленьким доходом;
- на проведение свадьбы.

Вывод: Самыми привлекательными являются клиенты без детей в статусе женат/замужем, берущие кредит на операции с недвижимостью. Данные категории клиентов самые многочисленные и самые ответственные.

Группой риска являются клиенты с одним или двумя детьми, не женатые или в гражданском браке, берущие кредит на операции с автомобилем.


